syntax = "proto3";

package knowledge_service;

// 知识库服务定义
service KnowledgeService {
  // 聊天接口
  rpc Chat(ChatRequest) returns (ChatResponse);
  
  // 多轮对话聊天接口
  rpc ChatConversation(ConversationChatRequest) returns (ChatResponse);
  
  // 创建新对话
  rpc CreateConversation(CreateConversationRequest) returns (ConversationResponse);
  
  // 获取对话历史
  rpc GetConversationHistory(ConversationHistoryRequest) returns (ConversationHistoryResponse);
  
  // 列出对话
  rpc ListConversations(ListConversationsRequest) returns (ListConversationsResponse);
  
  // 更新对话
  rpc UpdateConversation(UpdateConversationRequest) returns (ConversationResponse);
  
  // 删除对话
  rpc DeleteConversation(DeleteConversationRequest) returns (DeleteConversationResponse);
  
  // 反馈接口
  rpc SubmitFeedback(FeedbackRequest) returns (FeedbackResponse);
  
  // 获取反馈历史
  rpc GetFeedbackHistory(FeedbackHistoryRequest) returns (FeedbackHistoryResponse);
  
  // 获取系统统计信息
  rpc GetStats(StatsRequest) returns (StatsResponse);
  
  // 搜索文档
  rpc SearchDocuments(SearchRequest) returns (SearchResponse);
  
  // 健康检查
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
  
  // 验证邮箱（用于对话保存）
  rpc VerifyEmail(EmailVerificationRequest) returns (EmailVerificationResponse);
  
  // 带邮箱验证的对话聊天
  rpc ChatWithEmailVerification(EmailChatRequest) returns (ChatResponse);
}

// 聊天请求
message ChatRequest {
  string question = 1;
  bool use_feedback = 2; // 是否使用反馈优化
}

// 多轮对话聊天请求
message ConversationChatRequest {
  string question = 1;  // 用户问题
  string conversation_id = 2;  // 对话ID
  bool use_feedback = 3;  // 是否使用反馈
  bool use_reranking = 4;  // 是否使用重排序
  int32 top_k = 5;  // 返回的文档数量
  float similarity_threshold = 6;  // 相似度阈值
  SystemConfig system_config = 7;  // 系统配置
  int32 max_history_turns = 8;  // 最大历史轮次
}

// 创建对话请求
message CreateConversationRequest {
  string title = 1;  // 对话标题
  string user_id = 2;  // 用户ID
}

// 对话响应
message ConversationResponse {
  string conversation_id = 1;  // 对话ID
  string title = 2;  // 对话标题
  string user_id = 3;  // 用户ID
  string created_at = 4;  // 创建时间
  string updated_at = 5;  // 更新时间
  bool is_archived = 6;  // 是否归档
}

// 对话历史请求
message ConversationHistoryRequest {
  string conversation_id = 1;  // 对话ID
  int32 limit = 2;  // 限制数量
  int32 offset = 3;  // 偏移量
}

// 消息
message Message {
  string message_id = 1;  // 消息ID
  string conversation_id = 2;  // 对话ID
  string content = 3;  // 消息内容
  string role = 4;  // 角色（user/assistant）
  string created_at = 5;  // 创建时间
  repeated string source_documents = 6;  // 源文档
}

// 对话历史响应
message ConversationHistoryResponse {
  string conversation_id = 1;  // 对话ID
  string title = 2;  // 对话标题
  repeated Message messages = 3;  // 消息列表
  int32 total_count = 4;  // 总消息数
}

// 列出对话请求
message ListConversationsRequest {
  string user_id = 1;  // 用户ID
  int32 limit = 2;  // 限制数量
  int32 offset = 3;  // 偏移量
  bool include_archived = 4;  // 是否包含已归档
}

// 列出对话响应
message ListConversationsResponse {
  repeated ConversationResponse conversations = 1;  // 对话列表
  int32 total_count = 2;  // 总对话数
}

// 更新对话请求
message UpdateConversationRequest {
  string conversation_id = 1;  // 对话ID
  string title = 2;  // 对话标题
  bool is_archived = 3;  // 是否归档
}

// 删除对话请求
message DeleteConversationRequest {
  string conversation_id = 1;  // 对话ID
}

// 删除对话响应
message DeleteConversationResponse {
  bool success = 1;  // 是否成功
}

// 聊天响应
message ChatResponse {
  string question = 1;
  string original_answer = 2;
  string final_answer = 3;
  repeated SourceDocument source_documents = 4;
  FeedbackInfo feedback_info = 5;
  bool success = 6;
  string error_message = 7;
}

// 来源文档
message SourceDocument {
  string content = 1;
  string source = 2;
  map<string, string> metadata = 3;
}

// 反馈信息
message FeedbackInfo {
  bool is_improved = 1;
  double confidence_score = 2;
  int32 feedback_count = 3;
  repeated SimilarQuestion similar_questions = 4;
}

// 相似问题
message SimilarQuestion {
  string question = 1;
  double similarity_score = 2;
  string feedback_type = 3;
}

// 反馈请求
message FeedbackRequest {
  string question = 1;
  string original_answer = 2;
  string feedback_type = 3; // positive, negative, corrected
  string corrected_answer = 4; // 可选，仅当feedback_type为corrected时
  string feedback_text = 5; // 可选，补充说明
  repeated SourceDocument source_documents = 6;
}

// 反馈响应
message FeedbackResponse {
  int32 feedback_id = 1;
  bool success = 2;
  string error_message = 3;
}

// 反馈历史请求
message FeedbackHistoryRequest {
  string question = 1;
}

// 反馈历史响应
message FeedbackHistoryResponse {
  repeated FeedbackRecord records = 1;
  bool success = 2;
  string error_message = 3;
}

// 反馈记录
message FeedbackRecord {
  int32 id = 1;
  string question = 2;
  string original_answer = 3;
  string user_feedback = 4;
  string corrected_answer = 5;
  string feedback_text = 6;
  string timestamp = 7;
  string question_hash = 8;
  repeated SourceDocument source_documents = 9;
}

// 统计信息请求
message StatsRequest {
}

// 统计信息响应
message StatsResponse {
  KnowledgeBaseStats knowledge_base = 1;
  FeedbackStats feedback_system = 2;
  SystemConfig system_config = 3;
  bool success = 4;
  string error_message = 5;
}

// 知识库统计
message KnowledgeBaseStats {
  int32 total_documents = 1;
  int32 total_chunks = 2;
  string vector_store_path = 3;
}

// 反馈统计
message FeedbackStats {
  int32 total_feedback = 1;
  int32 positive_feedback = 2;
  int32 negative_feedback = 3;
  int32 corrected_feedback = 4;
  int32 improved_answers = 5;
  double satisfaction_rate = 6;
}

// 系统配置
message SystemConfig {
  bool feedback_learning_enabled = 1;
  double confidence_threshold = 2;
  double similarity_threshold = 3;
  string feedback_db_path = 4;
}

// 搜索请求
message SearchRequest {
  string query = 1;
  int32 k = 2; // 返回结果数量
}

// 搜索响应
message SearchResponse {
  repeated SearchResult results = 1;
  bool success = 2;
  string error_message = 3;
}

// 搜索结果
message SearchResult {
  string content = 1;
  double score = 2;
  map<string, string> metadata = 3;
}

// 健康检查请求
message HealthCheckRequest {
}

// 健康检查响应
message HealthCheckResponse {
  bool healthy = 1;
  string status = 2;
  string version = 3;
}

// 邮箱验证请求
message EmailVerificationRequest {
  string email = 1;  // 邮箱地址
}

// 邮箱验证响应
message EmailVerificationResponse {
  bool is_valid = 1;  // 邮箱是否有效
  string user_id = 2;  // 用户ID（基于邮箱生成）
  bool success = 3;  // 操作是否成功
  string error_message = 4;  // 错误信息
}

// 带邮箱验证的聊天请求
message EmailChatRequest {
  string email = 1;  // 邮箱地址
  string question = 2;  // 用户问题
  string conversation_id = 3;  // 对话ID（可选，如果为空则创建新对话）
  bool use_feedback = 4;  // 是否使用反馈
  bool use_reranking = 5;  // 是否使用重排序
  int32 top_k = 6;  // 返回的文档数量
  float similarity_threshold = 7;  // 相似度阈值
  SystemConfig system_config = 8;  // 系统配置
  int32 max_history_turns = 9;  // 最大历史轮次
  string conversation_title = 10;  // 对话标题（创建新对话时使用）
}